src/main.rs

```diff
@@ -412,66 +418,71 @@ fn main() {
 					}
 				}
 				cpu::TickResult::InvalidOpcode => {
-					break FailureReason::InvalidOpcode;
+					break Err(FailureReason::InvalidOpcode);
 				}
 			}
 
 			if cpu_state.pc == 0xFFFF {
-				break FailureReason::None;
+				break Ok(());
 			}
 
 			for addr in &global_config.crash_addresses {
 				if cpu_state.pc == *addr {
-					break 'tick FailureReason::Crash;
+					break 'tick Err(FailureReason::Crash);
 				}
 			}
 
 			for addr in &test.crash_addresses {
 				if cpu_state.pc == *addr {
-					break 'tick FailureReason::Crash;
+					break 'tick Err(FailureReason::Crash);
 				}
 			}
 
 			if cpu_state.cycles_elapsed >= global_config.timeout
 				|| cpu_state.cycles_elapsed >= test.timeout
 			{
-				break FailureReason::Timeout;
+				break Err(FailureReason::Timeout);
 			}
 		};
 
-		if failure_reason != FailureReason::None {
-			println!(
-				"\x1B[91m{}: {} failed\x1B[0m:\n{}\n{}",
-				rom_path,
-				test.name,
-				match failure_reason {
-					FailureReason::InvalidOpcode => "Invalid opcode",
-					FailureReason::Crash => "Crashed",
-					FailureReason::Timeout => "Timeout",
-					FailureReason::None => "",
-				},
-				cpu_state
-			);
-		} else if let Some(result) = &test.result {
-			match result.compare(&cpu_state) {
-				Ok(..) => {
+		match condition {
+			Err(failure_reason) => {
+				println!(
+					"\x1B[91m{}: {} failed\x1B[0m:\n{}\n{}",
+					rom_path,
+					test.name,
+					match failure_reason {
+						FailureReason::InvalidOpcode => "Invalid opcode",
+						FailureReason::Crash => "Crashed",
+						FailureReason::Timeout => "Timeout",
+					},
+					cpu_state
+				);
+			}
+			Ok(()) => {
+				let result = if let Some(result) = &test.result {
+					result
+				} else {
 					if cli.silent < SILENCE_PASSING {
 						println!("\x1B[92m{}: {} passed\x1B[0m", rom_path, test.name);
 					}
-					continue;
-				}
-				Err(msg) => {
-					print!(
-						"\x1B[91m{}: {} failed\x1B[0m:\n{}",
-						rom_path, test.name, msg
-					);
+					continue; // This continue skips failure handling.
+				};
+				match result.compare(&cpu_state) {
+					Ok(()) => {
+						if cli.silent < SILENCE_PASSING {
```

Comment:
Yeah, I think it would be beneficial to separate the rest of testing logic as well. I mostly focused on removing the redundant match arm for `FailureReason::None`.